dlib/cuda/cpu_dlib.cpp

```diff
@@ -1467,6 +1467,48 @@ namespace dlib
             }
         }
 
+    // ------------------------------------------------------------------------------------
+
+        void mish (
+            tensor& dest,
+            const tensor& src
+        )
+        {
+            const auto d = dest.host();
+            const auto s = src.host();
+            for (size_t i = 0; i < src.size(); ++i)
+                d[i] = s[i] * std::tanh(std::log(1+std::exp(s[i])));
+        }
+
+        void mish_gradient(
+            tensor& grad,
+            const tensor& dest,
+            const tensor& gradient_input
+        )
+        {
+            const auto g = grad.host();
+            const auto d = dest.host();
+            const auto in = gradient_input.host();
+            if (is_same_object(gradient_input, grad))
+            {
+                for (size_t i = 0; i < dest.size(); ++i)
+                {
+                    auto delta = 2*std::exp(d[i]) + std::exp(2*d[i]) + 2;
+                    auto omega = 4*(d[i] + 1) + 4*std::exp(2*d[i]) + std::exp(3*d[i]) + std::exp(d[i])*(4*d[i] + 6);
+                    g[i] = in[i]*std::exp(d[i])*delta/(omega*omega);
```

Comment:
This kind of thing is numerically unstable, in the same way that `log(1+exp(x))` is unstable. If the inputs are not small, exp() will produce inf which will then convert to NaN in the next lines, causing training to fail.  This kind of thing needs to be written in a way that doesn't cause values of `d` with modest values (e.g. a `d` of 5000) to yield NaN.

Take a look at the graph of this function: https://www.wolframalpha.com/input/?i=derivative+of+x*tanh%28log%281%2Bexp%28x%29%29%29. Note how the gradient saturates at 1 and 0.  Again, you can use if statements similar to how `dlib::log1pexp` handles this kind of thing to make the gradient computation numerically stable.